solr/core/src/java/org/apache/solr/schema/TextField.java

```diff
@@ -175,6 +179,26 @@ protected boolean supportsAnalyzers() {
     return true;
   }
 
+  @Override
+  public Query getPrefixQuery(QParser parser, SchemaField sf, String termStr) {
+    if ("".equals(termStr)) {
+      return getExistenceQuery(parser, sf);
+    }
+    Term term = new Term(sf.getName(), termStr);
+    MultiTermQuery query;
+    SolrParams lParams = parser == null ? null : parser.getLocalParams();
+    if (lParams == null || !lParams.getBool("forceAutomaton", false)) {
```

Comment:
QQ: for multivalued query and !query, we want DynamicComplementPrefixQuery ? 

And DynamicComplementPrefixQuery doesn't use `Automaton`?

Reply:
DynamicComplementPrefixQuery definitely doesn't use automaton -- that's the source of the efficiency (esp. for a common usecase for FS): very broad prefixes (covering a large number of terms).

If I understand the first question correctly: from the perspective of prefix query, TextField is essentially _always_ "multivalued", regardless of whether the fieldtype is configured as multivalued or not (as the field is tokenized, the _indexed_ values are always multivalued per doc). So basically for TextField, we use the same query implementation that we use for StrField, except that because it's always multivalued (in a sense), we can never actually take the optimization that confers the "DynamicComplement" part of the name `DynamicComplementPrefixQuery`.

Making the query itself multivalued would be done either as a non-prefix Terms query (which iirc does use Automaton, but this `getPrefixQuery()` method would not be consulted in that case), or would be a disjunction of independent prefix queries, in which case the multi-valuing happens at a different layer, and at this level we're just treating it as a regular prefix query. (Query negation is similar -- it happens at a higher level, so from the perspective of this method, we're just a regular prefix query).